lxd/network_allocations.go

```diff
@@ -0,0 +1,213 @@
+//      type: string
+//      example: default
+//    - in: query
+//      name: all-projects
+//      description: Retrieve entities from all projects
+//      type: boolean
+//  responses:
+//    "200":
+//      description: API endpoints
+//      schema:
+//        type: object
+//        properties:
+//          type:
+//            type: string
+//            description: Response type
+//            example: sync
+//          status:
+//            type: string
+//            description: Status description
+//            example: Success
+//          status_code:
+//            type: integer
+//            description: Status code
+//            example: 200
+//          metadata:
+//            type: array
+//            description: List of network allocations used by a consuming entity
+//            items:
+//              $ref: "#/definitions/NetworkAllocations"
+//    "403":
+//      $ref: "#/responses/Forbidden"
+//    "500":
+//      $ref: "#/responses/InternalServerError"
+func networkAllocationsGet(d *Daemon, r *http.Request) response.Response {
+	var err error
+	projectName := queryParam(r, "project")
+	allProjects := shared.IsTrue(queryParam(r, "all-projects"))
+
+	var projectNames []string
+	err = d.db.Cluster.Transaction(r.Context(), func(ctx context.Context, tx *db.ClusterTx) error {
+		// Figure out the projects to retrieve.
+		if !allProjects {
+			projectNames = []string{projectName}
+		} else {
+			// Get all project names if no specific project requested.
+			projectNames, err = dbCluster.GetProjectNames(ctx, tx.Tx())
+			if err != nil {
+				return fmt.Errorf("Failed loading projects: %w", err)
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	// Helper function to get the default CIDR address (/32 or /128 mask for ipv4 or ipv6 respectively)
+	getDefaultCIDRAddr := func(addr string) (string, error) {
+		ip := net.ParseIP(addr)
+		if ip.To4() != nil {
+			return fmt.Sprintf("%s/32", addr), nil
+		} else if ip.To16() != nil {
+			return fmt.Sprintf("%s/128", addr), nil
+		} else {
+			return "", fmt.Errorf("%s is not a valid IP address.", addr)
+		}
+	}
+
+	result := make([]api.NetworkAllocations, 0)
+
+	// Then, get all the networks, their network forwards and their network load balancers.
+	for _, projectName := range projectNames {
+		networkNames, err := d.db.Cluster.GetNetworks(projectName)
+		if err != nil {
+			return response.SmartError(err)
+		}
+
+		// Get all the networks, their attached instances, their network forwards and their network load balancers.
+		for _, networkName := range networkNames {
+			n, err := network.LoadByName(d.State(), projectName, networkName)
+			if err != nil {
+				return response.SmartError(err)
+			}
+
+			leases, err := n.Leases(projectName, clusterRequest.ClientTypeNormal)
+			if err != nil {
+				return response.SmartError(err)
+			}
+
+			instanceNetAllocs := make([]api.NetworkAllocations, 0)
+			instanceHwAddrs := make([]string, 0)
```

Comment:
Can you instead make instanceHwAddrs and networkHostnames maps of `api.NetworkAllocations`, then you lookup (or initialise if needed) and append to the `Addresses` field.